presto-main/src/main/java/com/facebook/presto/sql/planner/LiteralInterpreter.java

```diff
@@ -133,13 +134,12 @@ else if (value.equals(Double.POSITIVE_INFINITY)) {
             }
         }
 
-        if (type.equals(VARCHAR)) {
+        if (type instanceof VarcharType) {
             if (object instanceof Slice) {
-                return new StringLiteral(((Slice) object).toStringUtf8());
+                return new Cast(new StringLiteral(((Slice) object).toStringUtf8()), type.getTypeSignature().toString(), false, true);
```

Comment:
You need to compare using the UTF-8 length instead of String.length, since String.length is the count of shorts in UTF-16.